diff --git a/com/sun/el/parser/AstIdentifier.java b/com/sun/el/parser/AstIdentifier.java
index 87059d3cee247..51424c9825762 100644
--- a/com/sun/el/parser/AstIdentifier.java
+++ b/com/sun/el/parser/AstIdentifier.java
@@ -113,7 +113,28 @@ public final class AstIdentifier extends SimpleNode {
             }
         }
         ctx.setPropertyResolved(false);
-        Object ret = ctx.getELResolver().getValue(ctx, null, this.image);
+
+        Object ret;
+        /* Putting the Boolean into the ELContext is part of a performance
+         * optimisation for ScopedAttributeELResolver. When looking up "foo",
+         * the resolver can't differentiate between ${ foo } and ${ foo.bar }.
+         * This is important because the expensive class lookup only needs to
+         * be performed in the later case. This flag tells the resolver if the
+         * lookup can be skipped.
+         */
+        if (parent instanceof AstValue) {
+            ctx.putContext(this.getClass(), Boolean.FALSE);
+        } else {
+            ctx.putContext(this.getClass(), Boolean.TRUE);
+        }
+        try {
+            ret = ctx.getELResolver().getValue(ctx, null, this.image);
+        } finally {
+            // Always reset the flag to false so the optimisation is not applied
+            // inappropriately
+            ctx.putContext(this.getClass(), Boolean.FALSE);
+        }
+
         if (! ctx.isPropertyResolved()) {
             // Check if this is an imported static field
             if (ctx.getImportHandler() != null) {
@@ -237,3 +258,5 @@ public final class AstIdentifier extends SimpleNode {
         }
     }
 }
+
+/* @generated */
\ No newline at end of file
